src/java/org/apache/cassandra/index/sai/disk/PostingListRangeIterator.java

```diff
@@ -109,7 +111,9 @@ protected PrimaryKey computeNext()
             if (rowId == PostingList.END_OF_STREAM)
                 return endOfData();
 
-            return primaryKeyMap.primaryKeyFromRowId(rowId);
+            PrimaryKey pk = primaryKeyMap.primaryKeyFromRowId(rowId);
+            scoreCacheMapper.mapSSTableRowIdToPrimaryKey(rowId, pk);
```

Comment:
not obvious to me what's going on here -- looks like this adds a default score of -1, but I'm not sure why

Reply:
doubly confusing in that QueryContext has an `assert score >= 0` line

Reply:
I agree that it is not obvious, and may be bad design--I'll need to look closer. This code block is converting from row id to primary key. The -1 is a default that will force the downstream consumer of the cache to recompute the similarity value, which should only happen when there are updates to the vector's value in the row.

Reply:
To clarify, at a higher level, we have this mapping method because when we get the score earlier in the stream of rows, we only have a row id to associate the score with. Zhao wanted us to defer loading the primary key, and as a result we need to cache the row id to score mapping, which is per sstable. At a meta level, this is the part of the code where I wish an iterator could have associated metadata so we could just pass the score along and avoid this remapping completely.

Reply:
Further, this work will be disrupted by VSEARCH-14 because we shouldn't ever have cases of -1 because with the solution to VSEARCH-14, we'll be ordering materialized rows, which means that it should hopefully be a one to one mapping for the score to the primary key. It might even be possible to modify the output of the vector posting list to stream the score out with the row ordering.

I'll be taking a closer look at this later tonight.

Reply:
> Further, this work will be disrupted by VSEARCH-14 because we shouldn't ever have cases of -1

Given my current understanding of VSEARCH-14, this assertion is no longer true. We will still have a chance of multiple indexed vectors for the same column in a row across multiple sstables, which forces us to compute the vector similarity again when we have the materialized row after reading from disk.

In general, we use -1 as a value to indicate we must recompute the score. We assert that it is greater than 0 to ensure the assumption that -1 is a meaningful "null" value in our primitive float space. I'll revisit this with better documentation.

Reply:
> doubly confusing in that QueryContext has an `assert score >= 0` line

Oops, I finally found the place where this is doubly confusing. I agree. I'll fix it.